text/rust/0000-rustup.md

```diff
@@ -0,0 +1,43 @@
+
+## Detailed Explanation
+
+Right now we have two buildpacks: rust-dist and cargo install. The former provides the Rust installation and the latter uses Cargo to compile source code. This proposal would add a third buildpack called rustup. Users would pick to either use rust-dist or rustup. The default would be to use rustup, given that is what most users will expect and it's more flexible. We'll continue to retain rust-dist as this may be easier for some users in offline environments.
+
+## Rationale and Alternatives
+
+- Do nothing. We'll miss out on use cases we could service.
+- Update rust-dist to cover some of the gaps, such as having nightlies and being able to install additional toolkits. This would be technically possible but would require some extreme automation, given that nightlies come out, well, nightly (which would imply we'd do the same with rust-dist). It would also require more entries in buildpack.toml if we're to support more toolkits.
+
+## Implementation
+
+The idea is to have a build plan like this:
+
+1. The rustup buildpack will provide both rust and rustup.
+2. The rust-dist buildpack will not change and will continue to provide rust.
+3. The cargo-install buildpack will change and will now provide two buildplans. Both will provide and require cargo. The first will require rust. The second will require rust and rustup.
+
+The idea is that we have a buildplan which will allow for either rust-dist and cargo-install to be used together or we'll have a buildplan that allows for rustup and cargo-install to be used together. When the rustup buildpack runs, it will consume the rustup and rust requirements from the buildplan and so the rust-dist buildpack while it still runs will effectively do nothing (i.e. it only installs rust if the rust requirement has not been yet met). Optionally, a builder could remove rust-dist or rustup if the builder author had a preverence for one over the other.
+
+Since a user cannot manipulate the buildplan directly, we are exposing two ways that a user can control which method is used to install Rust.
+
+1. Through an environment variable. We will add `BP_RUSTUP_ENABLED` which the rustup buildpack will examine. If it's true, which is the default, then the buildpack will proceed to use Rustup to install Rust. If it's set to false, the buildpack will do nothing, which will allow for some other buildpack, presently that's just rust-dist, to install Rust.
```

Comment:
Does using rustup by default mean that offline builds will not work by default even if the rust-dist buildpack has a vendored copy of the desired version?

Reply:
It depends. Handling offline support for Rust could be done in a few different ways. Primarily `rustup` will pull downloads from a repository and by default that requires the Internet. You can host those binaries on your Intranet though, so you could just point `rustup` to your own server with `RUSTUP_DIST_SERVER` that doesn't require Internet and continue using `rustup`.

I think that would be the best way to support an offline environment because you can a.) still use `rustup` and b.) it allows full control of what versions are available to rustup, including nightly.

If you wanted the more traditional offline approach we support, where you can bundle rust into the image then you could disable `rustup` and use the rust-dist buildpack. I haven't tested to see what happens if `rustup` fails to install Rust but my guess is that the `rustup` CNB would fail and the build would then fail.

An option would be to not fail the build if `rustup` fails. If `rustup` exit's non-zero, we could log a warning and continue instead. So long as the `rustup` CNB doesn't consume the `rust` build plan entry, I'd expect `rust-dist` to run and supply rust.

Do you have a preference on how that behavior would work?